File: haskell/src/Solutions/Euler312.hs
module Solutions.Euler312 where
import Utils.NumberTheory(invertMod)
import Utils.Numeric(fastPowerMod)
import Math.NumberTheory.Moduli.Chinese(chineseRemainder)
import Data.Maybe(fromJust)

getB :: Integer -> Integer -> Integer
getB n m = (fromJust (invertMod 18 m) * (n - 27)) `mod` m
c3 = (8 * fastPowerMod 12 ac2 (13^8)) `mod` 13^8
ac2 = fromJust $ chineseRemainder [(b2, 13^7), (3, 12)]
b2 = getB (fastPowerMod 3 d2 (13^7)) (13^7)
d2 = fromJust $ chineseRemainder [(c2, 13^6), (0, 12)]
c2 = (8 * fastPowerMod 12 ac1 (13^6)) `mod` 13^6
ac1 = fromJust $ chineseRemainder [(b1, 13^5), (3, 12)]
b1 = getB (fastPowerMod 3 d1 (13^5)) (13^5)
d1 = fromJust $ chineseRemainder [(c1, 13^4), (0, 12)]
c1 = (8 * fastPowerMod 12 ac0 (13^4)) `mod` 13^4
ac0 = fromJust $ chineseRemainder [(b0, 13^3), (3, 12)]
b0 = getB (fastPowerMod 3 d0 (13^3)) (13^3)
d0 = 10000 `mod` (13^2 * 12)

euler312 :: IO String
euler312 = return $ show c3

File: rust/src/solutions/euler312.rs
use crate::utils::number_theory::{chinese_remainder::chinese, invert_mod::InvertMod};
use crate::utils::numeric::modpow::ModPow;

pub fn euler312() -> String {
    let get_b = |n: i64, m: i64| (18.invert_mod(&m).unwrap() * (n - 27)) % m;
    let p = |e| 13i64.pow(e);
    let d0 = 10_000 % (13 * 13 * 12);
    let b0 = get_b(3.modpow(&d0, &p(3)), p(3));
    let ac0 = chinese(b0, p(3), 3, 12).unwrap();
    let c1 = 8 * 12.modpow(&ac0, &p(4)) % p(4);
    let d1 = chinese(c1, p(4), 0, 12).unwrap();
    let b1 = get_b(3.modpow(&d1, &p(5)), p(5));
    let ac1 = chinese(b1, p(5), 3, 12).unwrap();
    let c2 = 8 * 12.modpow(&ac1, &p(6)) % p(6);
    let d2 = chinese(c2, p(6), 0, 12).unwrap();
    let b2 = get_b(3.modpow(&d2, &p(7)), p(7));
    let ac2 = chinese(b2, p(7), 3, 12).unwrap();
    let c3 = 8 * 12.modpow(&ac2, &p(8)) % p(8);
    c3.to_string()
}
